itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/internal/update/ThingUpdateOSGiTest.java

```diff
@@ -0,0 +1,365 @@
+                .withChannel(ChannelBuilder.create(channelUID0).withType(channelTypeOldUID).build())
+                .withChannel(ChannelBuilder.create(channelUID1).withType(channelTypeOldUID).build()).build();
+
+        managedThingProvider.add(thing);
+
+        Thing updatedThing = assertThing(thing, 3);
+        assertThat(updatedThing.getChannels(), hasSize(2));
+
+        Channel channel1 = updatedThing.getChannel("testChannel1");
+        assertChannel(channel1, channelTypeNewUID, "Test Label", null);
+
+        Channel channel2 = updatedThing.getChannel("testChannel2");
+        assertChannel(channel2, channelTypeOldUID, "TestLabel", null);
+    }
+
+    @Test
+    public void testOnlyMatchingInstructionsUpdate() {
+        registerThingType(MULTIPLE_CHANNEL_THING_TYPE_UID);
+
+        ChannelTypeUID channelTypeOldUID = new ChannelTypeUID(BINDING_ID, "testChannelOldTypeId");
+        registerChannelTypes(channelTypeOldUID);
+
+        ThingUID thingUID = new ThingUID(MULTIPLE_CHANNEL_THING_TYPE_UID, THING_ID);
+        ChannelUID channelUID0 = new ChannelUID(thingUID, "testChannel0");
+        ChannelUID channelUID1 = new ChannelUID(thingUID, "testChannel1");
+
+        Thing thing = ThingBuilder.create(MULTIPLE_CHANNEL_THING_TYPE_UID, thingUID)
+                .withChannel(ChannelBuilder.create(channelUID0).withType(channelTypeOldUID).build())
+                .withChannel(ChannelBuilder.create(channelUID1).withType(channelTypeOldUID).build())
+                .withProperty(PROPERTY_THING_TYPE_VERSION, "2").build();
+
+        managedThingProvider.add(thing);
+
+        Thing updatedThing = assertThing(thing, 3);
+        assertThat(updatedThing.getChannels(), hasSize(1));
+
+        Channel channel1 = updatedThing.getChannel("testChannel1");
+        assertChannel(channel1, channelTypeOldUID, null, null);
+    }
+
+    @Test
+    public void testNotModifiedIfHigherVersion() {
+        registerThingType(ADD_CHANNEL_THING_TYPE_UID);
+
+        ChannelTypeUID channelTypeUID = new ChannelTypeUID(BINDING_ID, "testChannelTypeId");
+        registerChannelTypes(channelTypeUID);
+
+        ThingUID thingUID = new ThingUID(ADD_CHANNEL_THING_TYPE_UID, THING_ID);
+        Thing thing = ThingBuilder.create(ADD_CHANNEL_THING_TYPE_UID, thingUID)
+                .withProperty(PROPERTY_THING_TYPE_VERSION, "1").build();
+        managedThingProvider.add(thing);
+
+        waitForAssert(() -> assertThat(thing.getStatus(), is(ThingStatus.ONLINE)));
+        assertThat(thingRegistry.get(thingUID), is(sameInstance(thing)));
+        assertThat(thing.getChannels(), is(emptyCollectionOf(Channel.class)));
+    }
+
+    private Thing assertThing(Thing oldThing, int expectedNewThingTypeVersion) {
+        ThingUID thingUID = oldThing.getUID();
+
+        waitForAssert(() -> {
+            Thing updatedThing = thingRegistry.get(thingUID);
+            assertThat(updatedThing, is(not(sameInstance(oldThing))));
```

Comment:
For `updatedThing` or `oldThing`? IMO Eclipse is wrong here, because `assertThat` allows `null` as actual value and `sameInstance` uses `==` for matching, which is also null-safe. 

Can you check if `assertThat(updatedThing, is(notNullValue()));` solves this issues?

Reply:
If you mean adding this line additionally, then the answer is no - the error remains.
But I've just found out that the issue is resolved by declaring `updatedThing` to be nullable:
```
            @Nullable Thing updatedThing = thingRegistry.get(thingUID);
            assertThat(updatedThing, is(not(sameInstance(oldThing))));
```